src/zenml/metadata_stores/mysql_metadata_store.py

```diff
@@ -37,11 +42,93 @@ def get_tfx_metadata_config(
         metadata_store_pb2.ConnectionConfig,
         metadata_store_pb2.MetadataStoreClientConfig,
     ]:
-        """Return tfx metadata config for mysql metadata store."""
-        return metadata.mysql_metadata_connection_config(
+        """Return tfx metadata config for MySQL metadata store."""
+        config = MySQLDatabaseConfig(
             host=self.host,
             port=self.port,
             database=self.database,
-            username=self.username,
-            password=self.password,
         )
+
+        secret = self._get_mysql_secret()
+
+        # Set the user
+        if self.user:
+            if secret and secret.user:
+                raise RuntimeError(
+                    f"Both the metadata store {self.name} and the secret "
+                    f"{self.secret} within your secrets manager define "
+                    f"a username `{self.user}` and `{secret.user}`. Please "
+                    f"make sure that you only use one."
+                )
+            else:
+                config.user = self.user
+        else:
+            if secret and secret.user:
+                config.user = self.user
+            else:
+                raise RuntimeError(
+                    "Your metadata store does not have a username. Please "
+                    "provide it either by defining it upon registration or "
+                    "through a MySQL secret."
+                )
+
+        # Set the password
+        if self.password:
+            if secret and secret.password:
+                raise RuntimeError(
+                    f"Both the metadata store {self.name} and the secret "
+                    f"{self.secret} within your secrets manager define "
+                    f"a password. Please make sure that you only use one."
+                )
+            else:
+                config.password = self.password
+        else:
+            if secret and secret.user:
+                config.password = self.password
```

Comment:
As far as I can see from some of the tfx examples, there are instances where the connection config is created without a password. Should this be an option?

Reply:
if the password is optional, then it can stay so and the runtime error is not needed. My suggestion also includes two other code changes that might be relevant though